File: go/singleton/singleton_1.go
/*
=====================================================================
🌟 Go Singleton Example 🌟
---------------------------------------------------------------------
📝 Explain: This package demonstrates the Singleton pattern in Go.

🚀 Purpose: Ensures a single instance of a map[string]string.
=====================================================================
*/

// =========================
//	IMPLEMENTATION
// =========================

package main

import (
	"fmt"
	"sync"
)

type singleton map[string]string

var (
	once     sync.Once
	instance singleton
)

// New returns the singleton instance
func New() singleton {
	once.Do(func() {
		instance = make(singleton)
	})
	return instance
}

//=========================
// 		USAGE
//=========================

func main() {
	// Get the singleton instance
	s := New()
	s["this"] = "that"

	// Get another reference to the singleton instance
	s2 := New()

	// This should print: This is that
	fmt.Println("This is", s2["this"])
}

File: python/singleton/singleton_1.py

"""
=====================================================================
🌟 Python Singleton Example 🌟
---------------------------------------------------------------------
📝 Explain: This is one of the ways to implement a singleton in Python.
In this method, we check if an instance of the class does not exist,
we create it, if it does exist, we do not create a new instance and 
return the same.

🚀 Purpose: Ensure only one instance of a class exists.
=====================================================================
"""


#=========================
#	  IMPLEMENTATION
#=========================

class Singleton:
    _instance = None 

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(Singleton, cls).__new__(cls)
        return cls._instance
    

#=========================
#	       USAGE
#=========================
    
A = Singleton() 
B = Singleton()   
print(id(A))
print(id(B))